### PR TITLE
Add pod spec validation

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Now that pod specs are first class task template targets we should add some preliminary validation upon registration since we can reason about the expected object structure.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/998

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by kumare3** [comment]
ohh this is awesome! - just that we will have to potentially have different ones for v1.Pod, v2.Pod (if that ever happens).

File: pkg/manager/impl/validation/task_validator.go [hunk 1]
```diff
@@ -43,6 +42,26 @@ func validateContainer(task core.TaskTemplate, taskConfig runtime.TaskResourceCo
...
+		return errors.NewFlyteAdminErrorf(codes.InvalidArgument, "K8sPod task type targets must specify a non-empty pod spec")
+	}
+	var podSpec v1.PodSpec
+	jsonObj, err := json.Marshal(task.GetK8SPod().PodSpec)
+	if err != nil {
+		return errors.NewFlyteAdminErrorf(codes.InvalidArgument, "Failed to json marshal K8sPod task type target with err: %v", err)
+	}
+	if err = json.Unmarshal(jsonObj, &podSpec); err != nil {
+		return errors.NewFlyteAdminErrorf(codes.InvalidArgument, "Failed to json unmarshal K8sPod task type target with err: %v", err)
+	}
```
Comment (EngHabu) [inline pkg/manager/impl/validation/task_validator.go:56]: Should we move this pattern somewhere? json marshal json unmarshal... maybe here: https://github.com/flyteorg/flytestdlib/blob/master/utils/marshal_utils.go but it's ok if it's within flyteadmin utils somewhere?

File: pkg/manager/impl/validation/task_validator.go [hunk 2]
```diff
@@ -43,6 +42,26 @@ func validateContainer(task core.TaskTemplate, taskConfig runtime.TaskResourceCo
...
+	for _, container := range podSpec.Containers {
+		if errs := validation.IsDNS1123Label(container.Name); len(errs) > 0 {
```
Comment (EngHabu) [inline pkg/manager/impl/validation/task_validator.go:58]: Why only validate the name? do we have a dependency on that later? or just trying to prevent issues as early as possible? 
Should that entire function move into the compiler now that Pods are first class citizens?
Also should we use a more elaborate [validation function](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L143) if we are trying to prevent runtime errors as much as possible?

File: pkg/manager/impl/validation/task_validator.go (hunk 1 above)
Comment (katrogan) [reply pkg/manager/impl/validation/task_validator.go:56]: @EngHabu done https://github.com/flyteorg/flytestdlib/pull/90, please review

File: pkg/manager/impl/validation/task_validator.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/validation/task_validator.go:58]: @EngHabu just trying to prevent issues as soon as possible 
good point re: the compiler, will close out this PR in favor of https://github.com/flyteorg/flytepropeller/pull/286

i looked into k8s pod spec validation functions but there wasn't anything servicable since the spec can be incomplete at the time of registration so I'm not sure if the above linked validation function is necessarily useful at the compilation stage

**Comment by katrogan** [comment]
closing in favor of https://github.com/flyteorg/flytepropeller/pull/286
